### PR TITLE
[BEAM-3971, BEAM-4284] Remove fromProto for Pipeline and PTransform translation.

### PR DESCRIPTION
**Pull request by lukecwik**
This prevents runners from attempting to convert a pipeline proto into SDK pipeline construct. All runners should use the proto directly and utilities that work on the proto representation since there is no expectation that runners will be able to rehydrate all aspects of SDK constructs.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Spark
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by bsidhom** [comment]
I suspect this will break the non-portable Flink translator because it forces a proto round trip. It may be sufficient to just remove that round trip, but it may be a much more involved fix.

**Comment by bsidhom** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
I removed the Flink round trip already.

**Comment by lukecwik** [comment]
Run Java Precommit

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Samza ValidatesRunner

**Comment by bsidhom** [comment]
The Jenkins tests results appear to have been swallowed. Do you know what happened?

**Comment by lukecwik** [comment]
There is a dev@ thread on it. It seems as though people are seeing multi-hour delays before seeing any status appear.

**Comment by lukecwik** [comment]
Run Java Precommit

**Comment by lukecwik** [comment]
Run Samza ValidatesRunner

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Samza ValidatesRunner

**Comment by lukecwik** [comment]
Here is the passing prior run of Flink VR: https://github.com/apache/beam/pull/5833

I typed the trigger phrase a couple of times to get these going so now the runs are just overwriting each other.

**Comment by bsidhom** [comment]
FYI, the correct link is https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle_PR/7/

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/CombineTranslationTest.java [hunk 1]
```diff
@@ -179,6 +185,12 @@ public void leaveCompositeTransform(Node node) {
     }
   }
 
+  private static Coder<?> getAccumulatorCoder(
+      CombinePayload payload, RehydratedComponents components) throws IOException {
```
Comment (bsidhom) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/CombineTranslationTest.java]: Not something to necessarily change here, but do we still need `RehydratedComponents` at this point? It depends how it's used in general, but I suspect that using the components may be sufficient (in most cases).

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java [hunk 2]
```diff
@@ -912,30 +922,81 @@ public void translateNode(
...
+                Collections.emptyList(),
+                new DoFnOperator.MultiOutputOutputManagerFactory<>(mainTag, outputCoder),
+                windowingStrategy,
+                new HashMap<>(), /* side-input mapping */
+                Collections.emptyList(), /* side inputs */
+                context.getPipelineOptions(),
+                inputKvCoder.getKeyCoder(),
+                keySelector);
+
+        // our operator excepts WindowedValue<KeyedWorkItem> while our input stream
```
Comment (bsidhom) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java]: "expects" (I realize you didn't add this comment here, but it shows up in the diff)

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java [hunk 3]
```diff
@@ -912,30 +922,81 @@ public void translateNode(
...
+                    keyedWorkItemStream.getParallelism());
+
+        rawFlinkTransform.setStateKeyType(keyedWorkItemStream.getKeyType());
+        rawFlinkTransform.setStateKeySelectors(keyedWorkItemStream.getKeySelector(), null);
+
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        SingleOutputStreamOperator<WindowedValue<KV<K, OutputT>>> outDataStream =
+            new SingleOutputStreamOperator(
+                keyedWorkItemStream.getExecutionEnvironment(),
+                rawFlinkTransform) {}; // we have to cheat around the ctor being protected
```
Comment (bsidhom) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java]: Wow.

Side note: GitHub is painfully naive about generating diffs for these types of refactor.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CombineTranslation.java [hunk 4]
```diff
@@ -141,96 +99,28 @@ public FunctionSpec translate(
...
+    GlobalCombineFn<?, ?, ?> combineFn = combine.getTransform().getFn();
+    try {
+      return RunnerApi.CombinePayload.newBuilder()
+          .setAccumulatorCoderId(
+              components.registerCoder(
+                  extractAccumulatorCoder(combineFn, (AppliedPTransform) combine)))
+          .setCombineFn(
+              SdkFunctionSpec.newBuilder()
+                  .setEnvironmentId(
+                      components.registerEnvironment(Environments.JAVA_SDK_HARNESS_ENVIRONMENT))
```
[the remark this answers is not in the document]
Comment (bsidhom) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CombineTranslation.java]: Never mind. Looks like Ankur's adding that in #5838.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/GroupByKeyTranslator.java [hunk 5]
```diff
@@ -128,16 +126,8 @@ public void translate(
...
-        combineFn =
-            (CombineFnBase.GlobalCombineFn<? super InputT, ?, OutputT>)
-                CombineTranslation.getCombineFn(appliedPTransform)
-                    .orElseThrow(() -> new IOException("CombineFn not found in node."));
-      } catch (IOException e) {
-        throw new RuntimeException(e);
-      }
-
+      final CombineFnBase.GlobalCombineFn<? super InputT, ?, OutputT> combineFn =
+          ((Combine.PerKey) transform).getFn();
```
[the remark this answers is not in the document]
Comment (lukecwik) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/GroupByKeyTranslator.java]: No, it doesn't.

I believe it was copying the same methodology Flink was doing which is why we found this code here.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java (hunk 2 above)
Comment (lukecwik) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java]: fixed here and elsewhere.

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/CombineTranslationTest.java (hunk 1 above)
Comment (lukecwik) [reply runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/CombineTranslationTest.java]: RehyrdatedComponents is still useful as an object cache for coders/windowing strategy/...

Still undecided what to do in this space.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CombineTranslation.java [hunk 6]
```diff
@@ -141,96 +99,28 @@ public FunctionSpec translate(
...
-    @Override
-    public SdkFunctionSpec getCombineFn() {
-      return payload.getCombineFn();
-    }
-
-    @Override
-    public Coder<?> getAccumulatorCoder() {
-      return accumulatorCoder;
+    } catch (CannotProvideCoderException e) {
+      throw new IllegalStateException(e);
```
[the remark this answers is not in the document]
Comment (lukecwik) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CombineTranslation.java]: fixed

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Samza ValidatesRunner

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/CombineTranslationTest.java (hunk 1 above)
Comment (bsidhom) [reply runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/CombineTranslationTest.java]: Ack

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/GroupByKeyTranslator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bsidhom) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/GroupByKeyTranslator.java]: Thanks for confirming.

**Comment by lukecwik** [comment]
Run Java PreCommit

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Samza ValidatesRunner

**Pull request merged**